### PR TITLE
Qml marketplace

### PR DESCRIPTION
**Pull request by roxanneskelly**
Implement the existing Marketplace in QML, including:
* Browsing
* Item viewing
* Purchasing
* Clara.IO integration

NOTE: This is not a marketplace redesign, although some visuals might have minor differences.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16738/HighFidelity-Beta-PR14775-cfaf01b.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16738/HighFidelity-Beta-PR14775-cfaf01b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16738/HighFidelity-Beta-Interface-PR14775-cfaf01b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16738/HighFidelity-Beta-PR14775-cfaf01b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16738/HighFidelity-Beta-Interface-PR14775-cfaf01b.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16761/HighFidelity-Beta-PR14775-6123c4e.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16761/HighFidelity-Beta-PR14775-6123c4e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16761/HighFidelity-Beta-Interface-PR14775-6123c4e.dmg

File: libraries/networking/src/AccountManager.cpp [hunk 1]
```diff
@@ -263,13 +264,14 @@ void AccountManager::sendRequest(const QString& path,
...
-                                  Q_ARG(QVariantMap, propertyMap));
+                                  Q_ARG(QVariantMap, propertyMap),
+                                  Q_ARG(QUrlQuery, query));
         return;
     }
 
     QNetworkAccessManager& networkAccessManager = NetworkAccessManager::getInstance();
 
-    QNetworkRequest networkRequest = createRequest(path, authType);
+    QNetworkRequest networkRequest = createRequest(path, authType, query);
```
Comment (howard-stearns) [inline libraries/networking/src/AccountManager.cpp]: wow, was this a bug that has been sitting there? 

I'm a bit concerned about this solution, adding this parameter to AccountManger::createRequest, as I think that class supposed to be a plug-in replacement for QNetworkAccessManager.  If we add this parameter to that createRequest, we're no longer compatible.

But that leaves the problem of what to do (assuming that I'm right that this is fixing a bug of an unhandled parameter).  Is that parameter ever actually used by anyone? I.e., is it a bug to not be handled, but it has never been noticed because no one supplies a separate query string?  If so, then can we drop the optional argument in sendQuery?  Of if we need it, can we handle it by adding that to the URL? 

I don't know what the answer is, I'm just .... uncomfortable.

File: interface/src/commerce/QmlMarketplace.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
+//
+//  QmlMarketplace.h
+//  interface/src/commerce
+//
+//  Guard for safe use of Marketplace by authorized QML.
```
Comment (howard-stearns) [inline interface/src/commerce/QmlMarketplace.h]: Why do we need this? E.g., why not just use `request`, which is what the commerce/discovery stuff does elsewhere from .js and .qml?  See use of scripts/modules/request.js and interface/resources/qml/hifi/RootHttpRequest.qml (of which the latter is admittedly quite convoluted).

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16775/HighFidelity-Beta-PR14775-dfac0d8.apk

File: interface/resources/qml/hifi/commerce/marketplace/MarketplaceItem.qml [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+        onMarketplaceItemLikeResult: {
+            if (result.status !== 'success') {
+                console.log("Failed to get Marketplace Categories", result.data.message);
+            } else {
+                root.liked = !root.liked;
+                root.likes = root.liked ? root.likes + 1 : root.likes - 1;
+            }
+        }
+    }
+    function getFormattedDate(timestamp) {
```
Comment (howard-stearns) [inline interface/resources/qml/hifi/commerce/marketplace/MarketplaceItem.qml]: If we're not trying to be "bug compatible" - http://doc.qt.io/qt-5/qml-qtqml-date.html

File: interface/resources/qml/hifi/commerce/marketplace/MarketplaceListItem.qml [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+
+                MouseArea {
+                    anchors {
+                        left: likesText.left
+                        right: heart.right
+                        top: parent.top
+                        bottom: parent.bottom
+                    }
+                    onClicked: {
+                        if(isLoggedIn) {
```
Comment (howard-stearns) [inline interface/resources/qml/hifi/commerce/marketplace/MarketplaceListItem.qml]: super trivial: coding standard has space between `if` and `(`.

File: interface/resources/qml/hifi/commerce/marketplace/MarketplaceListItem.qml [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+
+        Image {
+            id: itemImage
+            
+            anchors {
+                top: header.bottom
+                left: parent.left
+                right: parent.right
+            }
+            height: width*0.5625
```
Comment (howard-stearns) [inline interface/resources/qml/hifi/commerce/marketplace/MarketplaceListItem.qml]: coding standard whitespace around `*`

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16775/HighFidelity-Beta-PR14775-dfac0d8.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16775/HighFidelity-Beta-Interface-PR14775-dfac0d8.dmg

File: libraries/networking/src/AccountManager.cpp (hunk 1 above)
Comment (roxanneskelly) [reply libraries/networking/src/AccountManager.cpp]: Didn't see anywhere else in the code that actually uses that parameter.

I did try adding the query params to the URL, but the code ends up URL Escaping the URL path portion which is another issue.

Part of me thinks that it makes sense to have a way to pass query parameters to apis that generate calls via URL, as those do come up (my call) and will likely come up in the future.

File: interface/src/commerce/QmlMarketplace.h (hunk 2 above)
Comment (roxanneskelly) [reply interface/src/commerce/QmlMarketplace.h]: Do those handle authenticated calls?  (the like/unlike and marketplace/items stuff uses auth)

File: interface/resources/qml/hifi/commerce/marketplace/MarketplaceListItem.qml (hunk 4 above)
Comment (roxanneskelly) [reply interface/resources/qml/hifi/commerce/marketplace/MarketplaceListItem.qml]: fixed

File: interface/resources/qml/hifi/commerce/marketplace/MarketplaceListItem.qml (hunk 5 above)
Comment (roxanneskelly) [reply interface/resources/qml/hifi/commerce/marketplace/MarketplaceListItem.qml]: will fix

File: interface/src/commerce/QmlMarketplace.h (hunk 2 above)
Comment (howard-stearns) [reply interface/src/commerce/QmlMarketplace.h]: Yes, these authenticate using Interface login.

We have hooks in Interface that cause .js XMLHttpRequest to authenticate when talking to the metaverse (whether that metaverse is production or staging, or something else for that matter), but only when talking to the metaverse. E.g., you can't scoop up the auth by having a script phone home to some third party server so that the server can see the data and re-use it.

However, it just so happens that the way we inject that authentication is really really hard to do in the XMLHttpRequest seen by .qml. (Different Javascript engine, too.) So... we had to really jump through hoops to make RootHttpRequest.qml work. In fact, what it actually does under the hood is that it sends a message to Javascript! And then we have a script in the default scripts that looks for any such message, sends it through the .js request module, and then sends the result back to .qml. (That's why RootHttpRequest.qml has to be in the root.) Given that this is such a kludge, I think it's fair to say, "yeah, I don't want to use that."  The only problem though, is that we have this problem all over the place, and I don't want to build a custom authenticated-request handler for each one. RootHttpRequest.qml _is_ our generalized authenticated-request handler (or any request really), which all then use the same syntax for the various options, and the same callback signature.

Comment (howard-stearns) [reply interface/src/commerce/QmlMarketplace.h]: Come to think of it, I think maybe I should have used a hybrid of the two approaches in the first place: Implement the exact request.js module in C++, and then expose that to .js, .qml, and probably also use it in some of the places where we now have add-hoc interfaces with their own handler methods (because I'm not convince that that signal anti-pattern correctly handles overlapping requests).

I didn't think to do that earlier because I had started with the .js and didn't know how much of https://www.npmjs.com/package/request we were going to need, so I wanted to stay flexible in .js. But it's pretty stable now.

File: scripts/system/marketplaces/marketplaces.js [hunk 6]
```diff
@@ -560,6 +562,18 @@ var onQmlMessageReceived = function onQmlMessageReceived(message) {
...
+        ui.open(MARKETPLACE_CHECKOUT_QML_PATH);
+        ui.tablet.sendToQml({
+            method: 'updateCheckoutQMLItemID',
+            params: message
+        });
+        break;
+    case 'marketplace_marketplaces':
+        // This is the chooser between marketplaces. Only OUR markteplace
+        // requires/makes-use-of wallet, so doesn't go through openMarketplace bottleneck.
+        ui.open(MARKETPLACES_URL, MARKETPLACES_INJECT_SCRIPT_URL);
```
Comment (howard-stearns) [inline scripts/system/marketplaces/marketplaces.js]: It _looks_ like maybe opening the marketplace app will launch the chooser, which is still html+injection?

Ultimately, the idea is that all users would open Marketplace.qml, and then, if not on quest, that qml would have a button on the bottom that closes the qml and instead opens the clara html.  Is that saved for a second phase? As it stands right now, does this code open the old html for the top level marketplaces?

File: scripts/system/marketplaces/marketplaces.js [hunk 7]
```diff
@@ -699,7 +713,9 @@ var onTabletScreenChanged = function onTabletScreenChanged(type, url) {
     var onMarketplaceItemTesterScreenNow = (
         url.indexOf(MARKETPLACE_ITEM_TESTER_QML_PATH) !== -1 ||
         url === MARKETPLACE_ITEM_TESTER_QML_PATH);
```
Comment (howard-stearns) [inline scripts/system/marketplaces/marketplaces.js]: This (and new line 718) seems like a weird construction that would never be hit. 
For both of these, wouldn't ui.checkIsOpen('QML', oneOfTheURLConstants) work?  

(Where that first arg, type, used to mean something, but it looks like that's been modified to no longer be used. Bit rot.)

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16806/HighFidelity-Beta-PR14775-d3cffa5.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16806/HighFidelity-Beta-PR14775-d3cffa5.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16806/HighFidelity-Beta-Interface-PR14775-d3cffa5.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16806/HighFidelity-Beta-PR14775-d3cffa5.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16806/HighFidelity-Beta-Interface-PR14775-d3cffa5.exe

File: scripts/system/marketplaces/marketplaces.js (hunk 6 above)
Comment (roxanneskelly) [reply scripts/system/marketplaces/marketplaces.js]: Yes, the code currently just opens the old code for the marketplace list...shouldn't be any change in the functionality there.

File: scripts/system/marketplaces/marketplace.js [hunk 8]
```diff
@@ -1,117 +1,36 @@
...
-Script.scriptEnding.connect(function () {
-    browseExamplesButton.clicked.disconnect(onClick);
-    tablet.removeButton(browseExamplesButton);
-    marketplaceWindow.visibleChanged.disconnect(onMarketplaceWindowVisibilityChanged);
-});
+//
+// Run the functions.
+//
+startup();
+Script.scriptEnding.connect(shutdown);
```
[the remark this answers is not in the document]
Comment (roxanneskelly) [reply scripts/system/marketplaces/marketplace.js]: Honestly, I don't think marketplace.js is used anywhere, so I might want to just remove it.

**Comment by rickcarr044** [comment]
Currently testing PR14775

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16901/HighFidelity-Beta-PR14775-0fdbca8.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16901/HighFidelity-Beta-PR14775-0fdbca8.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16901/HighFidelity-Beta-Interface-PR14775-0fdbca8.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16901/HighFidelity-Beta-PR14775-0fdbca8.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16901/HighFidelity-Beta-Interface-PR14775-0fdbca8.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17005/HighFidelity-Beta-PR14775-39ad36a.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17005/HighFidelity-Beta-PR14775-39ad36a.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17005/HighFidelity-Beta-Interface-PR14775-39ad36a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17005/HighFidelity-Beta-PR14775-39ad36a.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17005/HighFidelity-Beta-Interface-PR14775-39ad36a.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17025/HighFidelity-Beta-PR14775-78aa178.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17026/HighFidelity-Beta-PR14775-db9b44a.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17026/HighFidelity-Beta-PR14775-db9b44a.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17026/HighFidelity-Beta-Interface-PR14775-db9b44a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17026/HighFidelity-Beta-PR14775-db9b44a.exe

**Comment by rickcarr044** [comment]
Currently testing PR14775

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17029/HighFidelity-Beta-PR14775-364a169.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17029/HighFidelity-Beta-PR14775-364a169.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17029/HighFidelity-Beta-Interface-PR14775-364a169.dmg

**Comment by rickcarr044** [comment]
Paused testing PR14775

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17029/HighFidelity-Beta-PR14775-364a169.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17029/HighFidelity-Beta-Interface-PR14775-364a169.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17047/HighFidelity-Beta-PR14775-d3bc536.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17047/HighFidelity-Beta-PR14775-d3bc536.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17047/HighFidelity-Beta-Interface-PR14775-d3bc536.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17047/HighFidelity-Beta-PR14775-d3bc536.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17047/HighFidelity-Beta-Interface-PR14775-d3bc536.exe

File: interface/resources/qml/hifi/commerce/marketplace/SortButton.qml [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+import Hifi 1.0 as Hifi
+import QtQuick 2.9
+import QtQuick.Controls 2.2
+import QtGraphicalEffects 1.0
+import stylesUit 1.0
+import controlsUit 1.0 as HifiControlsUit
+import "../../../controls" as HifiControls
+import "../common" as HifiCommerceCommon
+import "qrc:////qml//hifi//models" as HifiModels  // Absolute path so the same code works everywhere.
+import "../common/sendAsset"
```
Comment (howard-stearns) [inline interface/resources/qml/hifi/commerce/marketplace/SortButton.qml]: why is this needed? myabe same question for line 22?

File: scripts/system/marketplaces/marketplace.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (howard-stearns) [reply scripts/system/marketplaces/marketplace.js]: in the next life...
